rfd/xxxx-config-resource-docs.md

```diff
@@ -0,0 +1,554 @@
+function, the `multichecker` and `singlechecker` drivers run by exposing a
+`Main` function that handles all of the program's output and CLI arguments. It
+is not possible to initialize a data structure outside the `Main` function and
+assign values to it during the `Run` function, since `Main`
+[exits](https://cs.opensource.google/go/x/tools/+/refs/tags/v0.8.0:go/analysis/multichecker/multichecker.go;l=59)
+the current process after completing, leaving the program no chance to process
+the data structure.
+
+The `singlechecker` and `multichecker` packages run static analysis by importing
+from a `checker` package, which is internal to the
+`analysis` package. While we could copy this code and
+modify it for our purposes, the reference generator does not require the
+general-purpose static analysis capabilities of the `analysis` package.
+
+### Converting structs to `Resource`s
+
+In Go's `ast` package, a [`Decl`](https://pkg.go.dev/go/ast#Decl) is an
+interface representing a declaration. The logic for generating a `Resource`
+from a given type declaration depends on the nature of the type declaration. 
+
+Start by asserting that the `Decl` is an `*ast.GenDecl` with a `Specs`
+containing a `TypeSpec` with `Type: StructType`. From there, we can map the
+fields within [`ast.StructType`](https://pkg.go.dev/go/ast#StructType) to fields
+in a new `Resource`.
+
+Each `StructType` has a [`*FieldList`](https://pkg.go.dev/go/ast#FieldList) that
+we can use to obtain field data, including comments and tags.
+
+|`Resource` Field|How to generate|
+|---|---|
+|`Resource.SourcePath`|The `file` property of the underlying `sourceDecl`.|
+|`Resource.Description`|`GenDecl.Doc`|
+|`Resource.SectionName`|`TypeSpec.Name.Name`, with camelcase converted to separate words.|
+|`Fields[*].Name`|Process the `Field.Tag` in the `FieldList`, extracting the value of the `json` tag.|
+|`Fields[*].Description`|Each [`Field`](https://pkg.go.dev/go/ast#Field) in a `FieldList` has a `*CommentGroup` that we can use to extract the field's GoDoc. If a comment begins with the name of the field, we can remove the field name.|
+|`Fields[*].Type`|We can use `Field.Type` within the `FieldList` for this. Follow the rules for processing field types [below](#processing-field-types) |
+|`YAMLExample`|Follow the rules for processing field types [below](#processing-field-types).|
+
+If a root type named within the config file is not a struct, the generator will
+exit with an error message.
+
+### Processing field types
+
+#### Custom fields
+
+Entries in the reference for custom fields, e.g., `Labels`, don't lend
+themselves to automatic generation because they implement custom unmarshalers,
+and we can't rely on `json` struct tags to indicate the types of these fields.
+
+For simplicity, we can describe custom fields by hardcoding their descriptions
+and example YAML values. To do so, we can include an `examples` directory within
```

Comment:
In this case, would it make sense to add an example YAML document for a custom type within the type's GoDoc?

E.g., for [`Labels`](https://github.com/gravitational/teleport/blob/c91da46765953688d218dcba4a4ce7acf606639f/api/types/role.go#L1137-L1140):

```go
// Labels is a wrapper around map
// that can marshal and unmarshal itself
// from scalar and list values
// Example YAML:
// 'env': 'test'
// '*': '*'
// 'region': ['us-west-1', -eu-central-1]
// 'reg': ^us-west-1|eu-central-1$'
// ---
type Labels map[string]utils.Strings
```

The generator could then exit with an error if a GoDoc for a custom type does not include an example YAML document, which it interprets as everything between `Example YAML` and the following `---`. If it can't parse the example YAML document as YAML, it also exits with an error.

Reply:
That sounds fine to me, +1.